### PR TITLE
Fix nan check and unused external source arguments in debug mode

### PR DESCRIPTION
**Pull request by ksztenderski**
Signed-off-by: ksztenderski <ksztenderski@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:

Fixes value check in debug mode when `data == nan`. 
Fixes passing of arguments to external source in debug mode. Now it accepts arguments with any name, including unsupported by the standard version.

### Affected modules and functionalities:

`_PipelineDebug`, `_ExternalSourceDebug`

### Key points relevant for the review:

Keyword arguments specification for debug mode `external_source`.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by ksztenderski** [comment]
!build

File: dali/python/nvidia/dali/_debug_mode.py [hunk 1]
```diff
@@ -446,7 +454,8 @@ def run(self, inputs, kwargs):
             self._check_device_classification(
                 self._kwargs_classification[key].device, classification.device, 'Argument', key)
 
-            if not classification.is_batch and classification.data != self._init_args[key]:
+            if not classification.is_batch and classification.data != self._init_args[key] and \
+                    not (math.isnan(classification.data) and math.isnan(self._init_args[key])):
```
Comment (JanuszL) [inline dali/python/nvidia/dali/_debug_mode.py:458]: Can I ask for a test, please?

File: dali/python/nvidia/dali/_debug_mode.py [hunk 2]
```diff
@@ -141,8 +143,8 @@ class _ExternalSourceDebug:
     """Debug mode version of ExternalSource operator."""
 
     def __init__(
-            self, source=None, num_outputs=None, batch_size=-1, cycle=None, name=None, layout=None,
-            batch=None, batch_info=None):
+            self, source=None, num_outputs=None, batch_size=-1, cycle=None, name=None, device='cpu',
```
[the remark this answers is not in the document]
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py:146]: done

File: dali/python/nvidia/dali/_debug_mode.py (hunk 1 above)
Comment (ksztenderski) [reply dali/python/nvidia/dali/_debug_mode.py:458]: done

**Comment by ksztenderski** [comment]
!build

File: dali/test/python/test_pipeline_debug.py [hunk 3]
```diff
@@ -591,3 +592,60 @@ def test_variable_batch_size():
...
+        pipe.run()
+
+
+def test_nan_check():
+    err_msg = "Argument 'fdata' for operator 'constant' unexpectedly changed value from*"
+    for values in [[np.nan, 1], [1, np.nan]]:
+        yield raises(RuntimeError, glob=err_msg)(_test_nan_check), values
+
+    for values in [[1, 1], [np.nan, np.nan]]:
+        yield _test_nan_check, values
```
Comment (JanuszL) [inline dali/test/python/test_pipeline_debug.py:651]: Shouldn't it raise too?

File: dali/test/python/test_pipeline_debug.py [hunk 4]
```diff
@@ -591,3 +592,60 @@ def test_variable_batch_size():
...
+def _test_nan_check(values):
+    pipe = nan_check_pipeline(iter(values))
+    pipe.build()
+    for _ in range(2):
+        pipe.run()
+
+
+def test_nan_check():
+    err_msg = "Argument 'fdata' for operator 'constant' unexpectedly changed value from*"
+    for values in [[np.nan, 1], [1, np.nan]]:
```
Comment (JanuszL) [inline dali/test/python/test_pipeline_debug.py:647]: Can you add a case:
```[[1, np.nan], [1, np.nan]]```
It should not raise as I understand.

File: dali/test/python/test_pipeline_debug.py (hunk 3 above)
Comment (JanuszL) [reply dali/test/python/test_pipeline_debug.py:651]: Now I get it.

**Review by JanuszL — APPROVED** [review]

File: dali/python/nvidia/dali/external_source.py [hunk 5]
```diff
@@ -56,7 +56,7 @@ def _check_data_batch(data, batch_size, layout):
             raise RuntimeError("The layout '{}' cannot describe {}-dimensional data".format(layout, dim))
 
 
-def _prep_data_for_feed_input(data, batch_size, layout, device_id=None):
+def _prep_data_for_feed_input(data, batch_size, layout, device_id=-1):
```
Comment (szalpal) [inline dali/python/nvidia/dali/external_source.py:59]: AFAIK, `None` and `-1` have different meanings in the API. Are we sure, that this won't break existing CPU-only code?

Comment (ksztenderski) [reply dali/python/nvidia/dali/external_source.py:59]: OK, I guess for my usage it can remain `None` and I will just use `-1` where necessary.

**Review by szalpal — APPROVED** [review]

**Comment by ksztenderski** [comment]
!build

**Pull request merged**